### ISSUE TITLE
Compressed Integer Conversion

### ISSUE THREAD

**Issue by bmhm** [zchunk/zchunk#16]
About your  comment https://github.com/zchunk/zchunk/blob/c831e22497f08bcb909aff6a413c2325e777bb80/src/lib/compint.c#L61-L63 

Congratulations, you just reinvented LEB128 (https://en.wikipedia.org/wiki/LEB128), except you reversed the continue-flag!
Thus, you just made us implement our own reversed LEB128-logic :laughing:!

---

Actual answer to our comment: Yes, there is.

On that wiki page there is a pseudo-code, which will work just fine, except that you have to reverse the bitmask logic.

---

By the way, I think it is easier to use bitmasks (`FLAG_LAST_BYTE = 0b10000000`), because it really shows us the highest bit is a flag. If you write "128" and don't even assign a constant, everyone reading the code will wonder why there is this magic number.

---

Java code: 

```java
  public static int toUnsignedInt(final byte[] compressedUnsignedInt) {
    int result = 0b0;
    int shift = 0;

    for (int index = 0; index < compressedUnsignedInt.length; index++) {
      final byte b = compressedUnsignedInt[index];
      final byte leadingZero = (byte) (b & ~COMPRESSED_INT_LAST_BYTE_FLAG);

      result |= leadingZero << shift;
      shift += 7;
    }

    return BigInteger.valueOf(Integer.toUnsignedLong(result)).intValueExact();
  }
```

As you can see, I extract the bytes  from the file beforehand (it cannot be more than 5 bytes anyway for an integer). I can have the length of the byte array seperately, which is feasible in java.

**Comment by jdieter** [comment]
Thanks, I'd never even heard of LEB128.  It's unfortunate that I'm only finding out about it now, after the format is fixed, but it's probably not a huge deal.

You're right about the advantage of using a bitmask.  I'll take a look at refactoring that part of my code.

Just to note, integers can be more than five bytes on 64-bit architectures.

**Comment by bmhm** [comment]
Yes, I already changed it to long in my code and added some checks (10 bytes max). I think this is what you also use in your code (implicitly on 64bit systems).
Maybe the byte length of compressed integer should be fixed to max = 10 bytes. That's 2^64 - 1. This way, we would not need to fiddle with 32bit vs 64bit. I think in your code it is dependent on `sizeof(size_t)`, which will read only up to 5 bytes on 32bit systems.

**Comment by jdieter** [comment]
> Yes, I already changed it to long in my code and added some checks (10 bytes max). I think this is what you also use in your code (implicitly on 64bit systems).
> Maybe the byte length of compressed integer should be fixed to max = 10 bytes. That's 2^64 - 1. This way, we would not need to fiddle with 32bit vs 64bit. I think in your code it is dependent on `sizeof(size_t)`, which will read only up to 5 bytes on 32bit systems.

On a 32-bit system that isn't using 64-bit size_t's, zchunk should fail to read any sizes that require more than 32 bits.  But, there is a define that you can use on 32 bit to set size_t to 64 bits (I've temporarily forgotten what it is), which would allow zchunk to work with a significantly larger maximum chunk size.

In the real world, I wouldn't expect that anybody would ever need >4GB chunk sizes, but one never knows.  After all, nobody needs more than 640K, right? ;)

> Btw: why not reserve LEB128 for a v2.0 header? ;-)

We could, but I'm not sure what the point would be.  We will always need to support reading the v1 header, so it's not like we could throw away the current code.

**Comment by bmhm** [comment]
Agreed. So on a 64bit system, the maximum number of bytes will be 10. On a 32bit system, the maximum number of bytes will be 5 (`sizeof(size_t) == 4 bytes)` and then `4 * 8 / 7 + 1 == 5`).
With five you can represent `0xFF_FF_FF_FF_8F`, which is `2^64 - 1`. Yes, that will definitely suffice. 

For compatiblity, I still chose `long` in java (width = 64 bits).

You can see the code I am using here:
https://github.com/zchunk/zchunk-java/blob/master/compressedint/src/main/java/io/github/zchunk/compressedint/CompressedIntUtil.java

There are only two methods, compress and decompress.
